### PR TITLE
fix: extract edge values from AfterCreate resource data

### PR DESCRIPTION
**Pull request by akeemphilbert**
## Summary
- `resource_service.Create`/`Update` now pass `data` + `refProps` to `BuildResourceGraph` so reference properties land in the `@graph` edges node where `EdgeValue` can read them.
- `AddEdgeToGraph` is now idempotent, so the `Triple.Created` projection replay doesn't duplicate edges that `BuildResourceGraph` already wrote to `Resource.Created.Data`.
- Malformed edge values (non-map `@id`, unexpected types) now surface as errors instead of being silently overwritten.

## Root cause

`resource_service.go` was calling `BuildResourceGraph(strippedData, nil, ...)` — stripped input with `nil` refProps means no property is recognized as a reference, so no edges node is written. `AfterCreate` hooks like `enrollmentBehavior.seedAttendance` called `EdgeValue(r.Data(), ctx, "studentId")` and always got `""`, silently skipping invoice generation, attendance seeding, and enrollment task creation.

## Related

- Fixes wepala/weos-private-presets#8
- Preset-level test coverage: wepala/weos-private-presets PR (`fix/issue-8-enrollment-edge-values`)

## Test plan
- [x] Unit tests: `BuildResourceGraph`/`EdgeValue`/`FlattenGraph` round-trip (`application/triple_extraction_test.go`)
- [x] Unit tests: `AddEdgeToGraph` idempotency for single-ref, array, and malformed cases
- [x] Replay test: `buildStateFromTransaction` with `Resource.Created` + `Triple.Created` produces a single edge entry
- [x] E2E: `TestCreateResource_EdgesReadableAfterCreate` (`tests/e2e/tasks_test.go`) — project filter finds task after create
- [x] E2E: `TestUpdateResource_EdgesReconciled` — Update-time edge change reconciles triples correctly
- [x] E2E: `TestListTasks_ByProject` (pre-existing) still passes — projection FK column populated

## Follow-ups (not blocking)

- `enrollmentBehavior.AfterCreate` still returns `nil` on all error branches (identified by silent-failure-hunter review); an architectural change to propagate `AfterCreate` errors is the durable fix. Track as separate issue.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: application/resource_service.go [hunk 1]
```diff
@@ -242,14 +242,13 @@ func (s *resourceService) Create(
...
 	refProps := ExtractReferenceProperties(rt.Schema(), rt.Context())
 
-	// Strip reference properties from the data — resources are atomic.
-	// References are recorded as Triple events on the entity for atomic UoW commit.
-	strippedData, refs, err := ExtractAndStripReferences(data, refProps)
+	// Extract reference triples for atomic UoW commit alongside the entity.
+	_, refs, err := ExtractAndStripReferences(data, refProps)
 	if err != nil {
-		return nil, fmt.Errorf("failed to strip references: %w", err)
+		return nil, fmt.Errorf("failed to parse resource data: %w", err)
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply application/resource_service.go:250]: Fixed in 1159065. `Create`/`Update` now call the new `ExtractReferenceTriples` helper, which decodes once and emits triples without marshaling a stripped payload that the caller would discard. `ExtractAndStripReferences` is preserved for any caller that still needs the stripped output.

File: application/resource_service.go [hunk 2]
```diff
@@ -443,13 +442,13 @@ func (s *resourceService) Update(
 
 	refProps := ExtractReferenceProperties(rt.Schema(), rt.Context())
 
-	// Strip reference properties — resources are atomic.
-	strippedData, newRefs, err := ExtractAndStripReferences(data, refProps)
+	// Extract reference triples for atomic UoW commit alongside the entity.
+	_, newRefs, err := ExtractAndStripReferences(data, refProps)
 	if err != nil {
-		return nil, fmt.Errorf("failed to strip references: %w", err)
+		return nil, fmt.Errorf("failed to parse resource data: %w", err)
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply application/resource_service.go:452]: Fixed in 1159065 — same change as the `Create` path: switched to `ExtractReferenceTriples` to skip the marshal cycle.

File: application/resource_service.go [hunk 3]
```diff
@@ -242,14 +242,13 @@ func (s *resourceService) Create(
...
+	// Extract reference triples for atomic UoW commit alongside the entity.
+	_, refs, err := ExtractAndStripReferences(data, refProps)
 	if err != nil {
-		return nil, fmt.Errorf("failed to strip references: %w", err)
+		return nil, fmt.Errorf("failed to parse resource data: %w", err)
 	}
 
-	graphData, err := BuildResourceGraph(strippedData, nil, entityID, rt.Name(), rt.Context())
+	graphData, err := BuildResourceGraph(data, refProps, entityID, rt.Name(), rt.Context())
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply application/resource_service.go:254]: Fixed in 1159065. `BuildResourceGraph` now handles `[]any` of strings for reference properties, materializing each entry as `{"@id": ...}` and storing the array on the predicate. `extractTriplesFromEdgesNode` was extended to read both single-ref and array-ref shapes back. New unit test `TestBuildResourceGraph_ArrayRef` pins the behavior.

File: application/resource_service.go [hunk 4]
```diff
@@ -443,13 +442,13 @@ func (s *resourceService) Update(
...
+	// Extract reference triples for atomic UoW commit alongside the entity.
+	_, newRefs, err := ExtractAndStripReferences(data, refProps)
 	if err != nil {
-		return nil, fmt.Errorf("failed to strip references: %w", err)
+		return nil, fmt.Errorf("failed to parse resource data: %w", err)
 	}
 
-	graphData, err := BuildResourceGraph(strippedData, nil, entity.GetID(), rt.Name(), rt.Context())
+	graphData, err := BuildResourceGraph(data, refProps, entity.GetID(), rt.Name(), rt.Context())
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply application/resource_service.go:456]: Fixed in 1159065 — same change as the `Create` path: `BuildResourceGraph` materializes array-valued refs, so the `Update`-time `Resource.Updated.Data` contains them in the edges node without waiting for `Triple` replay.

File: application/triple_extraction.go [hunk 5]
```diff
@@ -218,20 +224,36 @@ func AddEdgeToGraph(
 		}
 		graphArr = append(graphArr, edgesNode)
 	} else {
-		// Edges node exists — add the edge (accumulate into array for multi-valued predicates).
+		// Edges node exists. Either the predicate is absent (add it), a single
+		// ref (dedupe or promote to array), or an array (dedupe or append).
 		edgesNode, ok := graphArr[1].(map[string]any)
 		if !ok {
 			edgesNode = map[string]any{"@id": subjectID}
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply application/triple_extraction.go]: Fixed in 1159065. `AddEdgeToGraph` now returns an error when `len(@graph) >= 2` and `@graph[1]` is not a `map[string]any`, instead of silently replacing it. Existing malformed-shape tests cover both the unexpected-type and malformed-array-entry cases.

File: application/triple_extraction_test.go [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+	}
+
+	// Add a second distinct student via triple replay.
+	afterAdd, err := AddEdgeToGraph(graph, "https://schema.org/participant", "stu-2", "enr-1")
+	if err != nil {
+		t.Fatalf("AddEdgeToGraph: %v", err)
+	}
+
+	var doc map[string]any
+	_ = json.Unmarshal(afterAdd, &doc)
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply application/triple_extraction_test.go]: Fixed in 1159065. Replaced both `_ = json.Unmarshal(...)` patterns in `TestAddEdgeToGraph_MultiValued` and `TestAddEdgeToGraph_Idempotent_ArrayBranch` with checked unmarshal calls that `t.Fatalf` on error.

File: tests/e2e/tasks_test.go [hunk 7]
```diff
@@ -117,6 +117,99 @@ func TestCreateTask_WithProjectReference(t *testing.T) {
...
+	if upResp.StatusCode != http.StatusOK {
+		t.Fatalf("update task: expected 200, got %d", upResp.StatusCode)
+	}
+	upResp.Body.Close()
+
+	// Task should now be listed under project2, not project1.
+	taskUnderProject := func(pid string) bool {
+		url := fmt.Sprintf("/api/task?_filter[project][eq]=%s", pid)
+		r := env.doRequest(t, "GET", url, "", "admin@weos.dev")
+		defer r.Body.Close()
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply tests/e2e/tasks_test.go]: Fixed in 1159065. Removed the redundant `defer r.Body.Close()` (since `readJSON` already does that) and added an explicit `r.StatusCode != http.StatusOK` check before parsing JSON, so a non-200 from the list endpoint produces a clear failure message instead of a JSON-decode error on the envelope.

File: web/admin/pages/resources/course-instance/[id]/index.vue [hunk 8]
```diff
@@ -0,0 +1,801 @@
...
+          isPrimaryBilling: true,
+        })
+      }
+    }
+
+    let created = 0
+    for (const student of allSelectedStudents.value) {
+      const enrollmentData: Record<string, any> = {
+        studentId: student.id,
+        courseInstanceId: id,
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply web/admin/pages/resources/course-instance/[id]/index.vue]: No longer applicable — the admin UI files have been moved out of this PR; the diff is now backend-only (`application/`, `tests/e2e/`).

File: web/admin/pages/index.vue [hunk 9]
```diff
@@ -18,31 +18,341 @@
...
+
+    <!-- Billing Summary -->
+    <a-row v-if="canAccessInvoices" :gutter="[16, 16]" style="margin-bottom: 24px">
+      <a-col :xs="24" :sm="8">
+        <a-card :bordered="false" size="small">
+          <a-statistic
+            title="Outstanding"
+            :value="totalOutstanding"
+            :precision="2"
+            :prefix="currency"
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply web/admin/pages/index.vue]: No longer applicable — the admin UI / Person handler changes have been moved out of this PR. The current diff is backend-only and matches the PR title/description.

File: application/triple_extraction.go [hunk 10]
```diff
@@ -447,11 +552,29 @@ func BuildResourceGraph(
...
+				refs := make([]any, 0, len(v))
+				for _, item := range v {
+					if s, ok := item.(string); ok && s != "" {
+						refs = append(refs, map[string]any{"@id": s})
+					}
+				}
+				if len(refs) > 0 {
+					edgesNode[predicateIRI] = refs
+					hasEdges = true
+				}
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply application/triple_extraction.go:589]: Fixed in 7f6778c. `EdgeValue` and `FlattenGraph` now handle the array-of-refs shape that `BuildResourceGraph` writes for array-valued reference properties:

- `EdgeValue` returns the first `@id` for the array case (preserves the single-string contract for existing scalar-ref callers).
- New `EdgeValues` returns every `@id` in order for callers that explicitly want the multi-valued result.
- `FlattenGraph` emits a `[]string` under the original property name when the edge is an array.
- A shared `collectEdgeIDs` helper centralizes the unwrap logic so all three readers stay consistent.

New tests: `TestEdgeValues_ArrayRef`, `TestFlattenGraph_ArrayRef`.

File: tests/e2e/tasks_test.go [hunk 11]
```diff
@@ -117,6 +117,105 @@ func TestCreateTask_WithProjectReference(t *testing.T) {
...
+
+	projectID := env.seedProjectForUser(t, "Edge Test Project", "admin@weos.dev")
+	taskBody := fmt.Sprintf(`{
+		"name": "Edge-Carrying Task",
+		"status": "open",
+		"project": %q
+	}`, projectID)
+	resp := env.doRequest(t, "POST", "/api/task", taskBody, "admin@weos.dev")
+	if resp.StatusCode != http.StatusCreated {
+		t.Fatalf("create task: expected 201, got %d", resp.StatusCode)
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply tests/e2e/tasks_test.go]: Fixed in 7f6778c. The `t.Fatalf` for the create-task non-201 now calls `readJSON(t, resp)` first so the body is consumed/closed and the error envelope is included in the failure message.

File: tests/e2e/tasks_test.go [hunk 12]
```diff
@@ -117,6 +117,105 @@ func TestCreateTask_WithProjectReference(t *testing.T) {
...
+	taskID := readEnvelopeData(t, resp)["id"].(string)
+
+	// The fix-sensitive path: filtering tasks by the project reference exercises
+	// the projection's FK column, which is populated from the edges node of the
+	// entity's @graph. Before the fix, this column stayed NULL because
+	// BuildResourceGraph produced a @graph with no edges node.
+	listURL := fmt.Sprintf("/api/task?_filter[project][eq]=%s", projectID)
+	listResp := env.doRequest(t, "GET", listURL, "", "admin@weos.dev")
+	if listResp.StatusCode != http.StatusOK {
+		t.Fatalf("list by project: expected 200, got %d", listResp.StatusCode)
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply tests/e2e/tasks_test.go]: Fixed in 7f6778c — same change as the create branch: the list-by-project non-200 now calls `readJSON(t, listResp)` before `t.Fatalf`, so the body is consumed/closed and the error envelope is logged.

File: tests/e2e/tasks_test.go [hunk 13]
```diff
@@ -117,6 +117,105 @@ func TestCreateTask_WithProjectReference(t *testing.T) {
...
+		t.Fatalf("create task: expected 201, got %d", resp.StatusCode)
+	}
+	taskID := readEnvelopeData(t, resp)["id"].(string)
+
+	// Re-point it to project2 via Update.
+	updateBody := fmt.Sprintf(`{"name":"Movable Task","status":"open","project":%q}`, project2)
+	upResp := env.doRequest(t, "PUT", "/api/task/"+taskID, updateBody, "admin@weos.dev")
+	if upResp.StatusCode != http.StatusOK {
+		t.Fatalf("update task: expected 200, got %d", upResp.StatusCode)
+	}
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply tests/e2e/tasks_test.go:192]: Fixed in 7f6778c. Both error branches in `TestUpdateResource_EdgesReconciled` (create-task and update-task) now call `readJSON` on the response first, closing the body and including the parsed envelope in the `t.Fatalf` message.

File: tests/e2e/tasks_test.go [hunk 14]
```diff
@@ -117,6 +117,105 @@ func TestCreateTask_WithProjectReference(t *testing.T) {
...
+	taskUnderProject := func(pid string) bool {
+		t.Helper()
+		url := fmt.Sprintf("/api/task?_filter[project][eq]=%s", pid)
+		r := env.doRequest(t, "GET", url, "", "admin@weos.dev")
+		// Note: readJSON closes r.Body, so we don't defer Close here.
+		// Assert status before parsing — a non-200 produces clearer output
+		// than a JSON-decode failure on an error envelope.
+		if r.StatusCode != http.StatusOK {
+			t.Fatalf("list by project %s: expected 200, got %d", pid, r.StatusCode)
+		}
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply tests/e2e/tasks_test.go:206]: Fixed in 7f6778c. Restructured the `taskUnderProject` helper so `readJSON(t, r)` runs unconditionally — it consumes/closes the body in both the success and failure paths, and the result map is included in the `t.Fatalf` message on non-200.

File: application/event_handlers_test.go [hunk 15]
```diff
@@ -8,6 +8,8 @@ import (
 
 	"weos/domain/entities"
 	"weos/domain/repositories"
+
+	"github.com/akeemphilbert/pericarp/pkg/eventsourcing/domain"
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply application/event_handlers_test.go:12]: Not applicable — the current grouping is what `goimports -local weos` (the formatter golangci-lint runs in this repo) produces. Convention here is stdlib → local (`weos/...`) → third-party (`github.com/...`), e.g. `application/resource_service.go:3-19`. Verified locally:

```
$ goimports -local weos -d application/event_handlers_test.go
$ golangci-lint fmt --diff ./application/...
```

Both produce no diff, so CI will not flag it. Moving the `github.com/` import above the `weos/` imports as suggested would actually break the convention.

File: application/triple_extraction_test.go [hunk 16]
```diff
@@ -0,0 +1,473 @@
...
+	// The edge should still be a single {"@id": "stu-1"} map, not an array of duplicates.
+	var doc map[string]any
+	if err := json.Unmarshal(afterReplay, &doc); err != nil {
+		t.Fatalf("unmarshal: %v", err)
+	}
+	graphArr := doc["@graph"].([]any)
+	if len(graphArr) < 2 {
+		t.Fatalf("expected @graph with edges node, got %d elements", len(graphArr))
+	}
+	edges := graphArr[1].(map[string]any)
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply application/triple_extraction_test.go]: Fixed in 0dc3d88. Switched the unchecked `doc["@graph"].([]any)` and `graphArr[1].(map[string]any)` assertions in `TestAddEdgeToGraph_Idempotent`, `TestAddEdgeToGraph_MultiValued`, and `TestAddEdgeToGraph_Idempotent_ArrayBranch` to the checked ok-form with `t.Fatalf`, so a shape regression produces an actionable error instead of a panic.

File: application/triple_extraction.go [hunk 17]
```diff
@@ -379,10 +486,19 @@ func FlattenGraph(graphData, ldContext json.RawMessage) json.RawMessage {
...
+				case 0:
+					// Skip — no usable @id values.
+				case 1:
+					if _, isArr := val.([]any); isArr {
+						// Preserve array shape even when only one entry survives.
+						flat[propName] = ids
+					} else {
+						flat[propName] = ids[0]
 					}
+				default:
```
[the remark this answers is not in the document]
Comment (akeemphilbert) [reply application/triple_extraction.go:501]: Good catch — fixed in 0dc3d88. `RemoveEdgeFromGraph` now keeps the array shape even when filtering shrinks it to one entry. The previous length-1 unwrap-to-scalar would have caused an array-valued reference property to silently flip between array and scalar based purely on event-replay history. Added `TestRemoveEdgeFromGraph_PreservesArrayShape` that exercises the shrink-to-one path and also asserts `FlattenGraph` keeps the array shape.
